README.md
# zp_drilling

CHANGELOG.md
## 0.0.0 (October 16, 2023)
  - Merge pull request #1 from vocegueda/feature/setup
  - Data uploaded to bucket
  - Data and Weights added to dvc
  - Initial commit



Code/hubconf.py
# YOLOv5 🚀 by Ultralytics, GPL-3.0 license
"""
PyTorch Hub models https://pytorch.org/hub/ultralytics_yolov5

Usage:
    import torch
    model = torch.hub.load('ultralytics/yolov5', 'yolov5s')  # official model
    model = torch.hub.load('ultralytics/yolov5:master', 'yolov5s')  # from branch
    model = torch.hub.load('ultralytics/yolov5', 'custom', 'yolov5s.pt')  # custom/local model
    model = torch.hub.load('.', 'custom', 'yolov5s.pt', source='local')  # local repo
"""

import torch


def _create(name, pretrained=True, channels=3, classes=80, autoshape=True, verbose=True, device=None):
    """Creates or loads a YOLOv5 model

    Arguments:
        name (str): model name 'yolov5s' or path 'path/to/best.pt'
        pretrained (bool): load pretrained weights into the model
        channels (int): number of input channels
        classes (int): number of model classes
        autoshape (bool): apply YOLOv5 .autoshape() wrapper to model
        verbose (bool): print all information to screen
        device (str, torch.device, None): device to use for model parameters

    Returns:
        YOLOv5 model
    """
    from pathlib import Path

    from yolov5.models.common import AutoShape, DetectMultiBackend
    from yolov5.models.experimental import attempt_load
    from yolov5.models.yolo import ClassificationModel, DetectionModel, SegmentationModel
    from yolov5.utils.downloads import attempt_download
    from yolov5.utils.general import LOGGER, check_requirements, intersect_dicts, logging
    from yolov5.utils.torch_utils import select_device

    if not verbose:
        LOGGER.setLevel(logging.WARNING)
    check_requirements(exclude=('opencv-python', 'tensorboard', 'thop'))
    name = Path(name)
    path = name.with_suffix('.pt') if name.suffix == '' and not name.is_dir() else name  # checkpoint path
    try:
        device = select_device(device)
        if pretrained and channels == 3 and classes == 80:
            try:
                model = DetectMultiBackend(path, device=device, fuse=autoshape)  # detection model
                if autoshape:
                    if model.pt and isinstance(model.model, ClassificationModel):
                        LOGGER.warning('WARNING ⚠️ YOLOv5 ClassificationModel is not yet AutoShape compatible. '
                                       'You must pass torch tensors in BCHW to this model, i.e. shape(1,3,224,224).')
                    elif model.pt and isinstance(model.model, SegmentationModel):
                        LOGGER.warning('WARNING ⚠️ YOLOv5 SegmentationModel is not yet AutoShape compatible. '
                                       'You will not be able to run inference with this model.')
                    else:
                        model = AutoShape(model)  # for file/URI/PIL/cv2/np inputs and NMS
            except Exception:
                model = attempt_load(path, device=device, fuse=False)  # arbitrary model
        else:
            cfg = list((Path(__file__).parent / 'models').rglob(f'{path.stem}.yaml'))[0]  # model.yaml path
            model = DetectionModel(cfg, channels, classes)  # create model
            if pretrained:
                ckpt = torch.load(attempt_download(path), map_location=device)  # load
                csd = ckpt['model'].float().state_dict()  # checkpoint state_dict as FP32
                csd = intersect_dicts(csd, model.state_dict(), exclude=['anchors'])  # intersect
                model.load_state_dict(csd, strict=False)  # load
                if len(ckpt['model'].names) == classes:
                    model.names = ckpt['model'].names  # set class names attribute
        if not verbose:
            LOGGER.setLevel(logging.INFO)  # reset to default
        return model.to(device)

    except Exception as e:
        help_url = 'https://github.com/ultralytics/yolov5/issues/36'
        s = f'{e}. Cache may be out of date, try `force_reload=True` or see {help_url} for help.'
        raise Exception(s) from e


def custom(path='path/to/model.pt', autoshape=True, _verbose=True, device=None):
    # YOLOv5 custom or local model
    return _create(path, autoshape=autoshape, verbose=_verbose, device=device)


def yolov5n(pretrained=True, channels=3, classes=80, autoshape=True, _verbose=True, device=None):
    # YOLOv5-nano model https://github.com/ultralytics/yolov5
    return _create('yolov5n', pretrained, channels, classes, autoshape, _verbose, device)


def yolov5s(pretrained=True, channels=3, classes=80, autoshape=True, _verbose=True, device=None):
    # YOLOv5-small model https://github.com/ultralytics/yolov5
    return _create('yolov5s', pretrained, channels, classes, autoshape, _verbose, device)


def yolov5m(pretrained=True, channels=3, classes=80, autoshape=True, _verbose=True, device=None):
    # YOLOv5-medium model https://github.com/ultralytics/yolov5
    return _create('yolov5m', pretrained, channels, classes, autoshape, _verbose, device)


def yolov5l(pretrained=True, channels=3, classes=80, autoshape=True, _verbose=True, device=None):
    # YOLOv5-large model https://github.com/ultralytics/yolov5
    return _create('yolov5l', pretrained, channels, classes, autoshape, _verbose, device)


def yolov5x(pretrained=True, channels=3, classes=80, autoshape=True, _verbose=True, device=None):
    # YOLOv5-xlarge model https://github.com/ultralytics/yolov5
    return _create('yolov5x', pretrained, channels, classes, autoshape, _verbose, device)


def yolov5n6(pretrained=True, channels=3, classes=80, autoshape=True, _verbose=True, device=None):
    # YOLOv5-nano-P6 model https://github.com/ultralytics/yolov5
    return _create('yolov5n6', pretrained, channels, classes, autoshape, _verbose, device)


def yolov5s6(pretrained=True, channels=3, classes=80, autoshape=True, _verbose=True, device=None):
    # YOLOv5-small-P6 model https://github.com/ultralytics/yolov5
    return _create('yolov5s6', pretrained, channels, classes, autoshape, _verbose, device)


def yolov5m6(pretrained=True, channels=3, classes=80, autoshape=True, _verbose=True, device=None):
    # YOLOv5-medium-P6 model https://github.com/ultralytics/yolov5
    return _create('yolov5m6', pretrained, channels, classes, autoshape, _verbose, device)


def yolov5l6(pretrained=True, channels=3, classes=80, autoshape=True, _verbose=True, device=None):
    # YOLOv5-large-P6 model https://github.com/ultralytics/yolov5
    return _create('yolov5l6', pretrained, channels, classes, autoshape, _verbose, device)


def yolov5x6(pretrained=True, channels=3, classes=80, autoshape=True, _verbose=True, device=None):
    # YOLOv5-xlarge-P6 model https://github.com/ultralytics/yolov5
    return _create('yolov5x6', pretrained, channels, classes, autoshape, _verbose, device)


if __name__ == '__main__':
    import argparse
    from pathlib import Path

    import numpy as np
    from PIL import Image

    from yolov5.utils.general import cv2, print_args

    # Argparser
    parser = argparse.ArgumentParser()
    parser.add_argument('--model', type=str, default='yolov5s', help='model name')
    opt = parser.parse_args()
    print_args(vars(opt))

    # Model
    model = _create(name=opt.model, pretrained=True, channels=3, classes=80, autoshape=True, verbose=True)
    # model = custom(path='path/to/model.pt')  # custom

    # Images
    imgs = [
        'data/images/zidane.jpg',  # filename
        Path('data/_images/zidane.jpg'),  # Path
        'https://ultralytics.com/images/zidane.jpg',  # URI
        cv2.imread('data/_images/bus.jpg')[:, :, ::-1],  # OpenCV
        Image.open('data/_images/bus.jpg'),  # PIL
        np.zeros((320, 640, 3))]  # numpy

    # Inference
    results = model(imgs, size=320)  # batched inference

    # Results
    results.print()
    results.save()


bump-version.sh
#!/bin/bash

# Thanks goes to @pete-otaqui for the initial gist:
# https://gist.github.com/pete-otaqui/4188238
#
# Original version modified by Marek Suscak
#
# works with a file called VERSION in the current directory,
# the contents of which should be a semantic version number
# such as "1.2.3" or even "1.2.3-beta+001.ab"

# this script will display the current version, automatically
# suggest a "minor" version update, and ask for input to use
# the suggestion, or a newly entered value.

# once the new version number is determined, the script will
# pull a list of changes from git history, prepend this to
# a file called CHANGELOG.md (under the title of the new version
# number), give user a chance to review and update the changelist
# manually if needed and create a GIT tag.

NOW="$(date +'%B %d, %Y')"
RED="\033[1;31m"
GREEN="\033[0;32m"
YELLOW="\033[1;33m"
BLUE="\033[1;34m"
PURPLE="\033[1;35m"
CYAN="\033[1;36m"
WHITE="\033[1;37m"
RESET="\033[0m"

LATEST_HASH=`git log --pretty=format:'%h' -n 1`

QUESTION_FLAG="${GREEN}?"
WARNING_FLAG="${YELLOW}!"
NOTICE_FLAG="${CYAN}❯"

ADJUSTMENTS_MSG="${QUESTION_FLAG} ${CYAN}Now you can make adjustments to ${WHITE}CHANGELOG.md${CYAN}. Then press enter to continue."
PUSHING_MSG="${NOTICE_FLAG} Pushing new version to the ${WHITE}origin${CYAN}..."

if [ -f VERSION ]; then
    BASE_STRING=`cat VERSION`
    BASE_LIST=(`echo $BASE_STRING | tr '.' ' '`)
    V_MAJOR=${BASE_LIST[0]}
    V_MINOR=${BASE_LIST[1]}
    V_PATCH=${BASE_LIST[2]}
    echo -e "${NOTICE_FLAG} Current version: ${WHITE}$BASE_STRING"
    echo -e "${NOTICE_FLAG} Latest commit hash: ${WHITE}$LATEST_HASH"
    V_MINOR=$((V_MINOR + 1))
    V_PATCH=0
    SUGGESTED_VERSION="$V_MAJOR.$V_MINOR.$V_PATCH"
    echo -ne "${QUESTION_FLAG} ${CYAN}Enter a version number [${WHITE}$SUGGESTED_VERSION${CYAN}]: "
    read INPUT_STRING
    if [ "$INPUT_STRING" = "" ]; then
        INPUT_STRING=$SUGGESTED_VERSION
    fi
    echo -e "${NOTICE_FLAG} Will set new version to be ${WHITE}$INPUT_STRING"
    echo $INPUT_STRING > VERSION
    echo "## $INPUT_STRING ($NOW)" > tmpfile
    git log --pretty=format:"  - %s" "v$BASE_STRING"...HEAD >> tmpfile
    echo "" >> tmpfile
    echo "" >> tmpfile
    cat CHANGELOG.md >> tmpfile
    mv tmpfile CHANGELOG.md
    echo -e "$ADJUSTMENTS_MSG"
    read
    echo -e "$PUSHING_MSG"
    git add CHANGELOG.md VERSION
    git commit -m "Bump version to ${INPUT_STRING}."
    git tag -a -m "Tag version ${INPUT_STRING}." "v$INPUT_STRING"
    git push origin --tags
else
    echo -e "${WARNING_FLAG} Could not find a VERSION file."
    echo -ne "${QUESTION_FLAG} ${CYAN}Do you want to create a version file and start from scratch? [${WHITE}y${CYAN}]: "
    read RESPONSE
    if [ "$RESPONSE" = "" ]; then RESPONSE="y"; fi
    if [ "$RESPONSE" = "Y" ]; then RESPONSE="y"; fi
    if [ "$RESPONSE" = "Yes" ]; then RESPONSE="y"; fi
    if [ "$RESPONSE" = "yes" ]; then RESPONSE="y"; fi
    if [ "$RESPONSE" = "YES" ]; then RESPONSE="y"; fi
    if [ "$RESPONSE" = "y" ]; then
        echo "0.0.0" > VERSION
        echo "## 0.0.0 ($NOW)" > CHANGELOG.md
        git log --pretty=format:"  - %s" >> CHANGELOG.md
        echo "" >> CHANGELOG.md
        echo "" >> CHANGELOG.md
        echo -e "$ADJUSTMENTS_MSG"
        read
        echo -e "$PUSHING_MSG"
        git add VERSION CHANGELOG.md
        git commit -m "Add VERSION and CHANGELOG.md files, Bump version to v0.0.0."
        git tag -a -m "Tag version 0.0.0." "v0.0.0"
        git push origin --tags
    fi
fi

echo -e "${NOTICE_FLAG} Finished."

Code/custom_dataset.py
import numpy as np
import cv2
from torch.utils.data import Dataset
from torch import get_rng_state, set_rng_state
from torchvision import transforms as tvt


class CustomDataset(Dataset):
    """
    This class is used to process oocytes images and masks
    """

    def __init__(self, x, y, n_classes=1,
                 transform_img=tvt.Compose([tvt.ToPILImage(),
                                            tvt.Resize((256, 256)),
                                            tvt.PILToTensor()]),
                 transform_mask=tvt.Compose([tvt.ToTensor(),
                                             tvt.Resize((256, 256), tvt.InterpolationMode.NEAREST)
                                             ])):
        self.x = x
        self.y = y
        self.n_classes = n_classes
        self.transform_img = transform_img
        self.transform_mask = transform_mask

    def __len__(self):
        return len(self.x)

    def __getitem__(self, ix):
        clahe = cv2.createCLAHE(clipLimit=2.0, tileGridSize=(8, 8))
        img = cv2.imread(self.x[ix], cv2.IMREAD_GRAYSCALE)
        img = clahe.apply(img).astype(np.float32)

        state = get_rng_state()
        img = self.transform_img(img)

        mask = np.load(self.y[ix])
        mask = (np.arange(self.n_classes) == mask[..., None]).astype(np.float32)
        set_rng_state(state)
        mask = self.transform_mask(mask)

        return img, mask


Code/compute_oocyte_poi.py
import numpy as np
import cv2
from skimage import measure
import segmentation_models_pytorch as smp
import torch
import torchvision.transforms as tvt
from PIL import Image
import argparse
import os
import sys
import json


def main():
    """
    Classes:
        0- Background
        1- Cytoplasm
        2- Perivitelline area
        3- Zona pellucida
        4- Polar body
    """

    # Sets the paths to detection and segmentation models
    det_weights_path = os.path.join(WEIGHT_DIR, DETECTION_WEIGHTS)
    seg_weights_path = os.path.join(WEIGHT_DIR, SEGMENTATION_WEIGHTS)

    # Gets the user parameters
    parser = argparse.ArgumentParser(description='script to compute some points of interest from a segmented oocyte')
    parser.add_argument('--input', type=str, required=True, help='path/to/image')
    parser.add_argument('--output', type=str, required=True, help='path/to/json_file')

    # Parses the user parameters
    args = parser.parse_args()
    img_path = args.input
    json_path = args.output

    dst_path = os.path.dirname(json_path)

    if dst_path.strip() == "":
        dst_path = './'

    # Performs some path verifications
    if not os.path.exists(det_weights_path):
        sys.exit('Error: path to "%s" does not exist' % det_weights_path)

    if not os.path.exists(seg_weights_path):
        sys.exit('Error: path to "%s" does not exist' % seg_weights_path)

    if not os.path.exists(img_path):
        sys.exit('Error: path to "%s" does not exist' % img_path)

    if not os.path.exists(dst_path):
        sys.exit('Error: directory "%s" does not exist' % dst_path)

    # Sets other parameters
    json_file = os.path.basename(json_path)
    base_name = json_file.split('.')[0]

    if json_file.strip() == "":
        sys.exit('Error: missing JSON filename')

    # Prepares the image transformations in a pipeline
    transform_img = tvt.Compose([tvt.ToPILImage(),
                                 tvt.Resize((HEIGHT, WIDTH), interpolation=tvt.InterpolationMode.BILINEAR),
                                 tvt.PILToTensor()])

    try:
        # Sets the device (CPU or GPU)
        device = torch.device('cuda:0') if torch.cuda.is_available() else torch.device('cpu')

        # Creates the detection model
        print('Loading the detection model...')
        det_model = torch.hub.load('.',
                                   'custom',
                                   path=det_weights_path,
                                   force_reload=True,
                                   source='local')

        # Initializes the detection model
        det_model.iou = 0.6  # IoU threshold
        det_model.conf = 0.6  # Confidence threshold
        det_model.max_det = 1  # Max number of detections allowed per frame
        print('Done\n')
        
        # Detection model predictions
        results = det_model(img_path)
        detections = results.xyxy[0].detach().cpu().numpy()

        # Checks if there are oocyte predictions
        if len(detections) > 0:
            # Extracts the bounding box (we are considering that only one object is detected)
            box = detections[0][:4]

            # Extracts the upper-left corner, width, and height from the bounding box
            x, y, w, h = round(box[0]), round(box[1]), round(box[2] - box[0]), round(box[3] - box[1])
        else:
            sys.exit('No oocyte detection')

        # Creates the segmentation model
        print('Loading the segmentation model...')

        seg_model = smp.Unet(
                encoder_name="resnet101",
                encoder_weights='imagenet',
                in_channels=N_CHANNELS,
                classes=N_CLASSES
                )

        # Puts the model in the device (CPU or GPU)
        seg_model = seg_model.to(device)

        # Loads the trained weights
        seg_model.load_state_dict(torch.load(seg_weights_path))

        # Sets the model to evaluation mode
        seg_model.eval()
        print('Done\n')

        # Reads the image
        color_img = cv2.imread(img_path)

        # Converts the image to grayscale
        img = cv2.cvtColor(color_img, cv2.COLOR_BGR2GRAY)

        # Crops the image using the predicted bounding box
        crop_img = img[y:y + h, x:x + w]
        crop_height = crop_img.shape[0]
        crop_width = crop_img.shape[1]

        # Applies the CLAHE algorithm to the crop
        clahe = cv2.createCLAHE(clipLimit=2.0, tileGridSize=(8, 8))
        crop_img = clahe.apply(crop_img).astype(np.float32)

        # Applies the transformations to the crop
        crop_img = transform_img(crop_img)

        # Puts the transformed tensor in the device (CPU or GPU)
        crop_img = torch.autograd.Variable(crop_img, requires_grad=False).to(device).unsqueeze(0)

        # Segmentation model predictions
        with torch.no_grad():
            prd = seg_model(crop_img)

        # Resizes the predicted mask
        prd = tvt.Resize((crop_height, crop_width), tvt.InterpolationMode.NEAREST)(prd[0])

        # Loads the predicted mask in the CPU as a numpy array
        seg = prd.data.cpu().detach().numpy()

        # Gets a binary mask for each channel
        seg = (seg > 0.5).astype(np.uint8)

        print('Calculating the points of interest on the oocyte and saving the results...')
        # Get the cytoplasm mask (channel 1)
        cy_mask = seg.copy()[1, :, :]
        cy_mask[cy_mask != 1] = 0
        cy_mask = cy_mask.astype(np.uint8)
        cy_size = np.count_nonzero(cy_mask)
        if cy_size == 0:
            sys.exit("Error: the class 'cytoplasm' was not found in the predicted mask")

        # Gets the cytoplasm centroid
        # A point of interest in the input image (not the crop)
        m = measure.moments(cy_mask)
        cx = round(m[0, 1] / m[0, 0])
        cy = round(m[1, 0] / m[0, 0])
        c = [x + cx, y + cy]

        # Gets the zp mask (channel 3)
        zp_mask = seg.copy()[3, :, :]
        zp_mask[zp_mask != 1] = 0
        zp_mask = zp_mask.astype(np.uint8)
        zp_size = np.count_nonzero(zp_mask)
        if zp_size == 0:
            sys.exit("Error: the class 'zona pellucida' was not found in the predicted mask")

        # Gets the coordinates of the zp mask
        coord = cv2.findNonZero(zp_mask)
        coord = np.squeeze(coord, axis=1)
        coord = coord[(coord[:, 0] > cx) & (coord[:, 1] == cy)]

        # Gets the zp mask inner point
        # A point of interest in the input image (not the crop)
        p1 = coord[0].tolist()
        p1[0] += x
        p1[1] += y

        # Gets the zp mask outer point
        # A point of interest in the input image (not the crop)
        p2 = coord[-1].tolist()
        p2[0] += x
        p2[1] += y

        print('cytoplasm_centroid =', c)
        print('zp_inner_point =', p1)
        print('zp_outer_point =', p2)

        # Initializes some data structures to store the results
        data = []
        results = {'data': []}

        d = {'cytoplasm_centroid': c,
             'zp_inner_point': p1,
             'zp_outer_point': p2,
             }

        data.append(d)
        results['data'] = data

        # Dumps the results to a JSON file
        with open(os.path.join(os.path.join(dst_path, base_name + '.json')), 'w') as fp:
            json.dump(results, fp, indent=4)

        # Graph parameters
        '''
        color = np.array([255, 0, 0], dtype='uint8')
        alpha = 0.4
        color_img = cv2.cvtColor(color_img, cv2.COLOR_BGR2RGB)
        crop_img = color_img[y:y + h, x:x + w]

        masked_img = np.where(zp_mask[..., None], color, crop_img)
        out = cv2.addWeighted(color_img[y:y + h, x:x + w], 1 - alpha, masked_img, alpha, 0)
        cv2.circle(out, (c[0] - x, c[1] - y), 5, (0, 255, 0), -1)
        cv2.circle(out, (p1[0] - x, p1[1] - y), 5, (0, 255, 0), -1)
        cv2.circle(out, (p2[0] - x, p2[1] - y), 5, (0, 255, 0), -1)
        out = Image.fromarray(out).convert('RGB')

        crop_img = Image.fromarray(crop_img).convert('RGB')
        '''

        # Plots the image with the computed points of interest
        color_img = cv2.cvtColor(color_img, cv2.COLOR_BGR2RGB)
        cv2.circle(color_img, (c[0], c[1]), 5, (0, 255, 0), -1)
        cv2.circle(color_img, (p1[0], p1[1]), 5, (0, 255, 0), -1)
        cv2.circle(color_img, (p2[0], p2[1]), 5, (0, 255, 0), -1)
        color_img = Image.fromarray(color_img)
        color_img.save(os.path.join(dst_path, base_name + '.jpg'))
        print('Done')

    except Exception as e:
        print('Exception: %s' % str(e))


if __name__ == '__main__':
    # Main parameters
    WIDTH = 256
    HEIGHT = 256
    N_CHANNELS = 1
    N_CLASSES = 5
    WEIGHT_DIR = 'weights'
    DETECTION_WEIGHTS = 'oocyte_det.pt'
    SEGMENTATION_WEIGHTS = 'oocyte_seg.pt'

    # Main function call
    main()


Code/train_segmenter.py
import numpy as np
from custom_dataset import CustomDataset
import torch
from torch.utils.data import DataLoader
from torchvision import transforms as tvt
import segmentation_models_pytorch as smp
import pandas as pd
import matplotlib.pyplot as plt
from glob import glob
import os
import sys
from tqdm import tqdm
from time import sleep


def compute_iou(outputs, labels):
    """
    Function to compute the iou score
    Args:
        outputs (tensor): predicted masks
        labels (tensor): ground-truth masks

    Returns:
        IoU score (float)
    """

    # True if value > 0.5, False otherwise
    outputs, labels = torch.sigmoid(outputs) > 0.5, labels > 0.5
    smooth = 1e-6

    # batch x num_classes x height x weight
    b, n, h, w = outputs.shape

    # Iterates over the tensor channels, except channel 0 (background)
    iou_list = []

    try:
        for i in range(1, n):
            # Extracts a mask per channel
            _out, _labs = outputs[:, i, :, :], labels[:, i, :, :]

            # Computes the intersection over axis 1 and 2
            intersection = (_out & _labs).float().sum(axis=(1, 2))

            # Computes the union over axis 1 and 2
            union = (_out | _labs).float().sum(axis=(1, 2))

            # Computes the IoU score and inserts the value into a list
            iou = (intersection + smooth) / (union + smooth)
            iou_list.append(iou.mean().item())
    except Exception as e:
        print('Exception: %e' % str(e))

    # Returns the mean value of the IoU scores
    return np.mean(iou_list)


def fit(model, device, dataloader, epochs=100, lr=1e-3):
    """
    Function to train the model and save the best and the last weights under a selected metric
    Args:
        model (torch model): NN model
        device (torch device): device (CPU or GPU)
        dataloader (torch dataloader): loads images and labels
        epochs (int): number of epochs for training
        lr (float): learning rate

    Returns:
        a dictionary with the history training
    """

    # Loss functions
    criterion1 = smp.losses.DiceLoss('binary', from_logits=True)
    criterion2 = smp.losses.SoftBCEWithLogitsLoss()

    # Optimizer
    # optimizer = torch.optim.AdamW(params=model.parameters(), lr=lr, weight_decay=1e-4, amsgrad=True)
    optimizer = torch.optim.Adam(params=model.parameters(), lr=lr, weight_decay=1e-4, amsgrad=True)

    # Learning rate scheduler
    scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(
        optimizer,
        mode='min',
        factor=0.5,
        patience=3,
        verbose=True,
        min_lr=1e-6
    )

    # Puts the model in a device
    model.to(device)

    # Training history dictionary
    hist = {'loss': [], 'iou': [], 'val_loss': [], 'val_iou': []}

    # Initial validation loss reference
    val_loss_ref = 1.0

    try:
        # Training loop
        for epoch in range(1, epochs + 1):
            # Loads images and labels
            bar = tqdm(dataloader['train'])

            # Initializes the loss and IoU values for training
            train_loss, train_iou = [], []

            # Sets the model to train mode
            model.train()

            for images, masks in bar:
                # Puts the images and the labels in the device
                images, masks = images.to(device), masks.to(device)

                # Sets the gradients to None
                optimizer.zero_grad()

                # Model predictions
                y_hat = model(images)

                # Computes the training loss
                loss = criterion1(y_hat, masks.float()) + criterion2(y_hat, masks.float())
                # loss = criterion1(y_hat, masks.float())

                # Performs the network backpropagation
                loss.backward()

                # Performs the parameter update
                optimizer.step()

                # Computes the training parameters for the current batch
                iou = compute_iou(y_hat, masks)
                train_loss.append(loss.item())
                train_iou.append(iou)
                bar.set_description("loss=%.5f iou=%.5f" % (float(np.mean(train_loss)), float(np.mean(train_iou))))
                sleep(0.1)

            # Stores the training parameters for the current epoch
            hist['loss'].append(np.mean(train_loss))
            hist['iou'].append(np.mean(train_iou))
            bar = tqdm(dataloader['val'])

            # Initializes the loss and IoU values for validation
            val_loss, val_iou = [], []

            # Sets the model to evaluation mode
            model.eval()

            # Validation loop
            with torch.no_grad():
                for images, masks in bar:
                    # Puts the images and the labels in the device
                    images, masks = images.to(device), masks.to(device)

                    # Model predictions
                    y_hat = model(images)

                    # Computes the validation loss
                    loss = criterion1(y_hat, masks.float()) + criterion2(y_hat, masks.float())
                    # loss = criterion1(y_hat, masks.float())

                    # Stores the validation parameters for the current batch
                    iou = compute_iou(y_hat, masks)
                    val_loss.append(loss.item())
                    val_iou.append(iou)
                    bar.set_description("val_loss=%.5f val_iou=%.5f" %
                                        (float(np.mean(val_loss)), float(np.mean(val_iou))))
                    sleep(0.1)

            # Stores the validation parameters for the current epoch
            hist['val_loss'].append(np.mean(val_loss))
            hist['val_iou'].append(np.mean(val_iou))

            # Prints the results at ending the epoch
            print('Epoch %d/%d loss=%.5f iou=%.5f val_loss=%.5f val_iou=%.5f' %
                  (epoch,
                   epochs,
                   float(np.mean(train_loss)),
                   float(np.mean(train_iou)),
                   float(np.mean(val_loss)),
                   float(np.mean(val_iou))))

            # Saves the best model
            if np.mean(val_loss) < val_loss_ref:
                print('Saving model...')
                torch.save(model.state_dict(), os.path.join(RESULTS_DIR, BEST_WEIGHT_FILE))
                val_loss_ref = np.mean(val_loss)

            # Applies the learning rate scheduler policy
            scheduler.step(np.mean(val_loss))
            print()
            sleep(0.1)

        # Saves the final model
        print('Saving model...')
        torch.save(model.state_dict(), os.path.join(RESULTS_DIR, LAST_WEIGHT_FILE))
    except Exception as e:
        print('Exception: %s' % e)

    return hist


def main():
    """
    Classes:
        0- Background
        1- Cytoplasm
        2- Perivitelline area
        3- Zona pellucida
        4- Polar body
    """

    # Creates a directory for the current experiment (if the directory exists, the experiment is aborted)
    try:
        if not os.path.exists(RESULTS_DIR):
            os.makedirs(RESULTS_DIR)
        else:
            sys.exit('Path to %s already exists. Please, assign another directory for the experiment results' %
                     RESULTS_DIR)
    except Exception as e:
        sys.exit('Exception: %s' % str(e))

    # Composes the image transformations
    transform_img = tvt.Compose([tvt.ToPILImage(),
                                 tvt.Resize((HEIGHT, WIDTH),
                                            interpolation=tvt.InterpolationMode.BILINEAR),
                                 # tvt.RandomRotation(ANGLE),
                                 tvt.RandomAffine(degrees=(-15, 15),
                                                  shear=(-15, 15),
                                                  interpolation=tvt.InterpolationMode.BILINEAR),
                                 # tvt.RandomVerticalFlip(p=0.5),
                                 tvt.PILToTensor()
                                 ])

    # Prepares the image transformations in a pipeline
    transform_mask = tvt.Compose([tvt.ToTensor(),
                                  tvt.Resize((HEIGHT, WIDTH), tvt.InterpolationMode.NEAREST),
                                  # tvt.RandomRotation(ANGLE),
                                  tvt.RandomAffine(degrees=(-15, 15),
                                                   shear=(-15, 15),
                                                   interpolation=tvt.InterpolationMode.NEAREST),
                                  # tvt.RandomVerticalFlip(p=0.5)
                                  ])

    # Gets the image/mask names
    img_names = sorted(glob(os.path.join(DATA_DIR, IMG_DIR, '*.jpg')))
    mask_names = sorted(glob(os.path.join(DATA_DIR, MASK_DIR, '*.npy')))

    # Gets the number of images and masks
    n_img = len(img_names)
    n_mask = len(mask_names)

    # Compares the images/masks sizes
    if n_img != n_mask:
        sys.exit('Error: the number of inputs and targets are different')

    # Sets the train data size
    train_size = round(n_img * TRAIN_DATA_FACTOR)

    # Gets the train/validation datasets
    dataset = {
        'train': CustomDataset(img_names[:train_size],
                               mask_names[:train_size],
                               N_CLASSES,
                               transform_img,
                               transform_mask),
        'val': CustomDataset(img_names[train_size:],
                             mask_names[train_size:],
                             N_CLASSES,
                             transform_img,
                             transform_mask)
    }

    # Configures the train/validation dataloaders
    dataloader = {
        'train': DataLoader(dataset['train'], batch_size=BATCH_SIZE, shuffle=True, pin_memory=True),
        'val': DataLoader(dataset['val'], batch_size=BATCH_SIZE, shuffle=False, pin_memory=True)
    }

    # Sets the device (CPU or GPU)
    device = 'cuda:0' if torch.cuda.is_available() else 'cpu'

    # Creates the NN model
    model = smp.Unet(
        encoder_name=ENCODER_NAME,
        encoder_weights='imagenet',
        in_channels=N_CHANNELS,
        classes=N_CLASSES,
        # activation='sigmoid'
    )

    # Trains the model
    try:
        hist = fit(model, device, dataloader, epochs=N_EPOCHS, lr=LR)
        df = pd.DataFrame(hist)
        df.to_csv(os.path.join(RESULTS_DIR, TRAIN_CSV_FILE), sep=',', index=False, encoding='latin1')
        title = 'Model: %s  Encoder: %s  Loss function: %s\n Input image format: %s' % \
                (MODEL_NAME, ENCODER_NAME, LOSS_FUNCTION_NAME, IMG_FORMAT)
        plt.tight_layout()
        df.plot(grid=True, title=title, xlabel='Epoch', ylabel='Loss / IoU score', figsize=(10, 8))
        # plt.show()
        plt.savefig(os.path.join(RESULTS_DIR, TRAIN_IMG_FILE))
    except Exception as e:
        sys.exit('Exception: %s' % str(e))


if __name__ == '__main__':
    # Main parameters
    MODEL_NAME = 'Unet'
    ENCODER_NAME = 'resnet101'
    LOSS_FUNCTION_NAME = 'Dice + SoftBCEWithLogits'
    IMG_FORMAT = 'grayscale'
    DATA_DIR = 'data'
    IMG_DIR = 'images'
    MASK_DIR = 'masks'
    WEIGHT_DIR = 'weights'
    BEST_WEIGHT_FILE = 'best.pt'
    LAST_WEIGHT_FILE = 'last.pt'
    RUN_DIR = 'runs'
    TRAIN_DIR = 'train'
    TEST_DIR = 'test'
    TRAIN_CSV_FILE = 'train.csv'
    TRAIN_IMG_FILE = 'train.jpg'
    WIDTH = 256  # image width
    HEIGHT = 256  # image height
    LR = 1e-3
    BATCH_SIZE = 10
    N_EPOCHS = 50
    N_CLASSES = 5
    N_CHANNELS = 1
    ANGLE = 15
    N_EXP = 1
    TRAIN_DATA_FACTOR = 0.8
    RESULTS_DIR = os.path.join(RUN_DIR, TRAIN_DIR, str(N_EXP))
    # SEED = 42
    # np.random.seed(seed)
    # torch.manual_seed(seed)
    # torch.cuda.manual_seed(seed)

    # Main function call
    main()
